app/src/main/java/io/github/trojan_gfw/igniter/TrojanService.java

```diff
@@ -74,18 +130,54 @@ public int onStartCommand(Intent intent, int flags, int startId) {
             b.addDnsServer("2001:4860:4860::8844");
         }
         pfd = b.establish();
+
+        if ((pfd == null) || !Tun2socks.setNonblock((long) pfd.getFd(), false)) {
```

Comment:
does the blocking mode affects throughput

Reply:
IMO, since a loop in Java thread is used in network IO, we have to use block mode. I prefer nonblock mode too, but don't know how to implement it properly.

Reply:
If you really want it to be blocking, try https://developer.android.com/reference/android/net/VpnService.Builder.html#setBlocking(boolean)